### PR TITLE
Validate rule [oas-descr]

### PR DESCRIPTION
**Pull request by MDoevenspeck**

### CODE REVIEW

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+function void violationTitleShouldNotBePresentForNonObjectTypesInInlineSchemas(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "The title property of a an inline schema MUST NOT be used for non-object types.",
+    "Title present in inline schema with a non-object type.",
+     schema);
+}
+
+rule "A summary should be present for each operation"
+  when
+    $operation: OperationDefinition(/model[summary == null])
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Summary allowed to be empty?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+function Boolean isInlineSchemaInAllOf(SchemaDefinition parent, SchemaDefinition schemaDefinition) {
+    if (parent == null || parent.getModel() == null) return false;
+
+    List<Schema> allOfSchemas = parent.getModel().getAllOf();
+    if (allOfSchemas == null) return false;
+
+    Schema schema = schemaDefinition.getModel();
+    return allOfSchemas.contains(schema);
+}
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: function not needed. SchemaDefinition.isHighLevelSchema will give the same result

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  then
+    violationSummaryShouldBePresentForEachOperation(oas, $operation);
+end
+
+rule "A title should be present for inline object schemas (not required in an AllOf / Not composition)"
+  when
+    $schema: SchemaDefinition(definitionType == DefinitionType.INLINE,
+    /model[type == SchemaType.OBJECT && title == null],
+    $parent: parent#SchemaDefinition,
+    !isInlineSchemaInAllOf($parent, this))
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Can be replaced with 'isHighLevelSchema'

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import io.github.belgif.rest.guide.validator.core.model.OpenApiDefinition.DefinitionType;
+import java.util.List
+import java.lang.String
+
+global io.github.belgif.rest.guide.validator.core.ViolationReport oas;
+
+function void violationSummaryShouldBePresentForEachOperation(ViolationReport oas, OperationDefinition operation){
+    oas.addViolation("[oas-descr]",
+    "The summary property of an operation SHOULD BE present for each operation.",
+    "Summary property not present.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: The message is (optionally) used to give details on this specific violation for this rule. So should be different for each invocation / dynamically put there.

In this case I think it's better to leave it out. Description (which should be the same for each set of violations that is grouped together) gives enough context in this case.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  when
+    $operation: OperationDefinition(/model[summary == null])
+  then
+    violationSummaryShouldBePresentForEachOperation(oas, $operation);
+end
+
+rule "A title should be present for inline object schemas (not required in an AllOf / Not composition)"
+  when
+    $schema: SchemaDefinition(definitionType == DefinitionType.INLINE,
+    /model[type == SchemaType.OBJECT && title == null],
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: title allowed to be empty?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+end
+
+rule "A title should be present for inline object schemas (not required in an AllOf / Not composition)"
+  when
+    $schema: SchemaDefinition(definitionType == DefinitionType.INLINE,
+    /model[type == SchemaType.OBJECT && title == null],
+    $parent: parent#SchemaDefinition,
+    !isInlineSchemaInAllOf($parent, this))
+  then
+    violationTitleShouldBePresentForInlineObjectSchemas(oas, $schema);
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl:68]: What should happen if a title is present on an inline schema within an allOf? Allowed?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    oas.addViolation("[oas-descr]",
+    "The summary property of an operation SHOULD BE present for each operation.",
+    "Summary property not present.",
+    operation);
+}
+
+function void violationTitleShouldBePresentForInlineObjectSchemas(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "A title property of an inline object schema SHOULD BE present (not required in an AllOf / Not composition)",
+    "Title property not present.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Same as for the other violation -> message is not necessary here

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+function void violationSummaryShouldBePresentForEachOperation(ViolationReport oas, OperationDefinition operation){
+    oas.addViolation("[oas-descr]",
+    "The summary property of an operation SHOULD BE present for each operation.",
+    "Summary property not present.",
+    operation);
+}
+
+function void violationTitleShouldBePresentForInlineObjectSchemas(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "A title property of an inline object schema SHOULD BE present (not required in an AllOf / Not composition)",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: For readability I propose:
'For inline schemas of type 'object' a title must be present.'

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    if (allOfSchemas == null) return false;
+
+    Schema schema = schemaDefinition.getModel();
+    return allOfSchemas.contains(schema);
+}
+
+function void violationTitleShouldNotBePresentOrMatchSchemaName(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "The title property of a component schema MUST NOT be used or MUST MATCH schema name.",
+    "Title present in component schema and doesn't match schema name.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: This would be a good spot to actually use the message and display something like:
'Title << ... >> found but schema name was << ... >>'

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+     schema);
+}
+
+function void violationTitleShouldNotBePresentForNonObjectTypesInInlineSchemas(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "The title property of a an inline schema MUST NOT be used for non-object types.",
+    "Title present in inline schema with a non-object type.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Same as for other violations. No need to use the message

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+end
+
+rule "A title should either be absent or match the component schema name"
+    when
+        $schema: SchemaDefinition(definitionType == DefinitionType.TOP_LEVEL,
+        $identifier: identifier#String,
+        /model[title != null && !$identifier.equals(title)]
+        )
+    then
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl:85]: Could be something like:
$schema: SchemaDefinition(definitionType == DefinitionType.TOP_LEVEL,
        $title: model.title#String,
$title != null,
        identifier != $title)

Then you can pass $title to the violation message

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    void testOperationSummaryPresent() {
+        assertNoViolations(callRules("summaryPresent.yaml"));
+    }
+
+    @Test
+    void testOperationSummaryNotPresent() {
+        ViolationReport violationReport = callRules("summaryNotPresent.yaml");
+        assertErrorCount(1, violationReport);
+        assertEquals("Summary property not present.", violationReport.getViolations().get(0).getMessage());
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Assertion not needed, assertErrorCount should be enough in this case.

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
+package io.github.belgif.rest.guide.validator.rules.oas;
+
+import io.github.belgif.rest.guide.validator.core.ViolationReport;
+import io.github.belgif.rest.guide.validator.rules.AbstractOasRuleTest;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class GeneralDocumentationTest extends AbstractOasRuleTest {
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: remove public

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        ViolationReport violationReport = callRules("titlePresentInCompentSchema.yaml");
+        assertErrorCount(1, violationReport);
+        assertEquals("Title present in component schema and doesn't match schema name.", violationReport.getViolations().get(0).getMessage());
+    }
+
+    @Test
+    void testComponentSchemaTitlePresentInInlineArraySchema() {
+        ViolationReport violationReport = callRules("titlePresentInInlineArraySchema.yaml");
+        assertErrorCount(1, violationReport);
+        assertEquals("Title present in inline schema with a non-object type.", violationReport.getViolations().get(0).getMessage());
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Should not assert on message

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    void testInlineSchemaTitlePresent() {
+        assertNoViolations(callRules("inlineSchemaTitlePresent.yaml"));
+    }
+
+    @Test
+    void testInlineSchemaTitleNotPresent() {
+        ViolationReport violationReport = callRules("inlineSchemaTitleNotPresent.yaml");
+        assertErrorCount(1, violationReport);
+        assertEquals("Title property not present.", violationReport.getViolations().get(0).getMessage());
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Should not assert on message

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    @Test
+    void testComponentSchemaTitleMatchesSchemaName() {
+        assertNoViolations(callRules("titleMatchesSchemaName.yaml"));
+    }
+
+    @Test
+    void testComponentSchemaTitlePresent() {
+        ViolationReport violationReport = callRules("titlePresentInCompentSchema.yaml");
+        assertErrorCount(1, violationReport);
+        assertEquals("Title present in component schema and doesn't match schema name.", violationReport.getViolations().get(0).getMessage());
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Do not assert on message

File: rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/generalDocumentation/titleNotPresent.yaml
Comment (smals-mavh) [inline rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/generalDocumentation/titleNotPresent.yaml:19]: Inline schema not flagged because of assumption that parent is of type SchemaDefinition

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    $operation: OperationDefinition(/model[summary == null])
+  then
+    violationSummaryShouldBePresentForEachOperation(oas, $operation);
+end
+
+rule "A title should be present for inline object schemas (not required in an AllOf / Not composition)"
+  when
+    $schema: SchemaDefinition(definitionType == DefinitionType.INLINE,
+    /model[type == SchemaType.OBJECT && title == null],
+    $parent: parent#SchemaDefinition,
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: You cannot assume that the parent is of type SchemaDefinition

File: rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/freeTextDocumentation/inlineSchemaTitleNotPresent.yaml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    get:
+      operationId: operationId
+      summary: operation summary
+      responses:
+        default:
+          content:
+            application/json:
+              schema:
+                type: object
+                description: my description
```
Comment (smals-mavh) [inline rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/freeTextDocumentation/inlineSchemaTitleNotPresent.yaml:16]: No schemaTitle, same for other tests

**Review by smals-mavh — CHANGES_REQUESTED** [review]
@MDoevenspeck ,

Overall good work, but you missed some edge cases.
Also the distinction between the violation description / message isn't used correctly (you can check in the code comments in the violation object).

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl (hunk 8 above)
Comment (smals-mavh) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Or should be, to verify -> recommended / mandatory?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl (hunk 3 above)
Comment (MDoevenspeck) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Is this correct? isHighLevelSchema does not distinguish between all,any or one of. 
Will {!isHighLevelSchema()} create positive matches on any and one of?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl (hunk 6 above)
Comment (MDoevenspeck) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl:68]: From the issue: in an inline subschema of an allOf or not composition, it isn't necessary to have a title for codegen. 
That  is why this condition is added to the rule: !isInlineSchemaInAllOf($parent, this)

File: rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/generalDocumentation/titleNotPresent.yaml
Comment (MDoevenspeck) [reply rules/src/test/resources/io/github/belgif/rest/guide/validator/rules/oas/generalDocumentation/titleNotPresent.yaml:19]: Not sure which inline schema you are talking about. Title should be present for type object, not for other types.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl (hunk 17 above)
Comment (MDoevenspeck) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Replaced logic, parent var is not used anymore

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl (hunk 6 above)
Comment (MDoevenspeck) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl:68]: After talking to Peter, it is allowed, name is not used in codegen

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+import io.github.belgif.rest.guide.validator.core.model.OpenApiDefinition;
+import io.github.belgif.rest.guide.validator.core.model.OpenApiDefinition.DefinitionType;
+import java.util.List
+import java.lang.String
+
+global io.github.belgif.rest.guide.validator.core.ViolationReport oas;
+
+function void violationSummaryShouldBePresentForEachOperation(ViolationReport oas, OperationDefinition operation){
+    oas.addViolation("[oas-descr]",
+    "The summary property of an operation SHOULD BE present for each operation.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Should be -> so violationLevel recommended.
@pvdbosch please verify, there are more rules that have SHOULD in the violation string but are still mandatory. Maybe leave it like this, and create a separate issue to review all violation levels?

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    operation);
+}
+
+function void violationTitleNamesShouldBeUpperCamelCase(ViolationReport oas, String title){
+   oas.addViolation("[oas-descr]",
+      "Title names SHOULD use UpperCamelCase notation. For abbreviations as well, all letters except the first one should be lowercased.",
+      title);
+}
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Either move this rule / violation to Rule-ComponentNamesShouldBeUpperCamelCase.drl or merge that rule with this file. (@pvdbosch any opinions?) (You can combine the violation functions probably)
You should also put the actual component (SchemaDefinition) in the violation (like for the other violation functions), otherwise lineNumbers and jsonpointers cannot be calculated.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+function void violationTitleShouldNotBePresentOrMatchSchemaName(ViolationReport oas, SchemaDefinition schema,String title){
+    oas.addViolation("[oas-descr]",
+    "The title property of a component schema MUST NOT be used or MUST MATCH schema name.",
+    "'Title << "+ title + " >> found, but schema name was << "+ schema.getIdentifier() +" >>'");
+}
+
+function void violationTitleShouldNotBePresentForNonObjectTypesInInlineSchemas(ViolationReport oas, SchemaDefinition schema){
+    oas.addViolation("[oas-descr]",
+    "The title property of a an inline schema MUST NOT be used for non-object types.",
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Typo: 'a an'

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+      !isSchemaInAllOrAnyOf(this)
+      )
+  then
+    violationTitleShouldBePresentForInlineObjectSchemas(oas, $schema);
+end
+
+rule "A title names should be UpperCamelCase"
+  when
+    $schema: SchemaDefinition($title: /model/title#String)
+    $title: String(this != null && !this.trim.empty && !ApiFunctions.isUpperCamelCase(this))
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: Maybe add code comment that explains that this rule should indeed only fire if title != null and that the null check is done in the other rules.

Technically I suppose you don't have to check the null and empty, because if /model/title == null, the rule won't continue further.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+function Boolean isSchemaInAllOrAnyOf(SchemaDefinition schemaDefinition) {
+    return schemaDefinition.getJsonPointer().toString().matches("^.*/(all|any)Of/\\d*?$");
+}
+
+function void violationTitleShouldNotBePresentOrMatchSchemaName(ViolationReport oas, SchemaDefinition schema,String title){
+    oas.addViolation("[oas-descr]",
+    "The title property of a component schema MUST NOT be used or MUST MATCH schema name.",
+    "'Title << "+ title + " >> found, but schema name was << "+ schema.getIdentifier() +" >>'");
+}
```
Comment (smals-mavh) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl:52]: This method now creates a violation with ruleName, description and message, but does not include the actual schema used to calculate lineNumbers and jsonPointer.
This should be added!

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Test
+    void testOperationSummaryPresent() {
+        assertNoViolations(callRules("summaryPresent.yaml"));
+    }
+
+    @Test
+    void testOperationSummaryNotPresent() {
+        ViolationReport violationReport = callRules("summaryNotPresent.yaml");
+        assertErrorCount(3, violationReport);
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Can be combined into assertErrorCount(3, callRules("summaryNotPresent.yaml"));

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/FreeTextDocumentationTest.java [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    void testInlineSchemaTitlePresent() {
+        assertNoViolations(callRules("inlineSchemaTitlePresent.yaml"));
+    }
+
+    @Test
+    void testInlineSchemaTitleNotPresent() {
+        ViolationReport violationReport = callRules("inlineSchemaTitleNotPresent.yaml");
+        assertErrorCount(3, violationReport);
+    }
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/FreeTextDocumentationTest.java:32]: Lines can be combined

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+    @Test
+    void testComponentSchemaTitleMatchesSchemaName() {
+        assertNoViolations(callRules("titleMatchesSchemaName.yaml"));
+    }
+
+    @Test
+    void testComponentSchemaTitlePresent() {
+        ViolationReport violationReport = callRules("titlePresentInCompentSchema.yaml");
+        assertErrorCount(1, violationReport);
+    }
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Lines can be combined

Also the name of the test is misleading. It tests if a violation is triggered if the component name != title

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/FreeTextDocumentationTest.java [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+    void testComponentSchemaTitlePresent() {
+        ViolationReport violationReport = callRules("titlePresentInCompentSchema.yaml");
+        assertErrorCount(1, violationReport);
+    }
+
+    @Test
+    void testComponentSchemaTitlePresentInInlineArraySchema() {
+        ViolationReport violationReport = callRules("titlePresentInInlineArraySchema.yaml");
+        assertErrorCount(1, violationReport);
+    }
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/FreeTextDocumentationTest.java]: Lines can be combined
Also bit misleading name for the test. I suppose you want to test if a title in a schema with type != object triggers a violation.
Maybe also add (in the same file) a property of type string and number with title.
Just for future us, if we have to revisit it / as a regression test.

File: rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java [hunk 28]
```diff
@@ -0,0 +1,61 @@
+package io.github.belgif.rest.guide.validator.rules.oas;
+
+import io.github.belgif.rest.guide.validator.core.ViolationReport;
+import io.github.belgif.rest.guide.validator.rules.AbstractOasRuleTest;
+import org.junit.jupiter.api.Test;
+
+class GeneralDocumentationTest extends AbstractOasRuleTest {
```
Comment (smals-mavh) [inline rules/src/test/java/io/github/belgif/rest/guide/validator/rules/oas/GeneralDocumentationTest.java]: Does not have a testcase for the UpperCamelCase rule.

**Review by smals-mavh — CHANGES_REQUESTED** [review]

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl (hunk 19 above)
Comment (pvdbosch) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: agreed, created #199

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl (hunk 20 above)
Comment (pvdbosch) [reply rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-GeneralDocumentation.drl]: I'd rather keep them separate, to have a single drl file for each rule in the REST guide, even though there may be some older drls not doing so yet.

Best to rename the file to Rule-FreeTextDocumentation , to align with the rule name in the guide

Alternatively we could rename all drl files to the REST guide rule code `oas-descr`, but they might be too short to be self-explanatory.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+    oas.addViolation("[oas-descr]",
+    "The summary property SHOULD be present for each operation.",
+    null,
+    operation,
+    ViolationLevel.RECOMMENDED);
+}
+
+function void violationTitleNamesShouldBeUpperCamelCase(ViolationReport oas, SchemaDefinition schema){
+   oas.addViolation("[oas-descr]",
+      "Title names SHOULD use UpperCamelCase notation. For abbreviations as well, all letters except the first one should be lowercased.",
```
Comment (pvdbosch) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl]: nitpick: 
-->  The title property of a schema SHOULD use UpperCamelCase notation.

It isn't enforced on other definitions.

File: rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+    when
+        $schema: SchemaDefinition(definitionType == DefinitionType.TOP_LEVEL,
+        $title: model.title#String,
+        $title != null,
+        identifier != $title)
+    then
+        violationTitleShouldNotBePresentOrMatchSchemaName(oas, $schema, $title);
+end
+
+rule "A title property in an inline schema should not be present for non-object types"
```
Comment (pvdbosch) [inline rules/src/main/resources/io/github/belgif/rest/guide/validator/rules/oas/Rule-FreeTextDocumentation.drl]: I don't think this is part of the belgif rule, so shouldn't be enforced

File: integrationtest/src/it/externalReferencesSucceeds/openapi.yaml [hunk 31]
```diff
@@ -332,6 +335,7 @@ paths:
     get:
       description: get logo metadata
       operationId: getLogoMetaData
+      summary: get logo metadata
```
Comment (pvdbosch) [inline integrationtest/src/it/externalReferencesSucceeds/openapi.yaml:333]: we'll have to remember to update the [rest-guide demo](https://github.com/belgif/rest-guide-example) API as well (+ implementation), bc this test was based on it.

It makes more sense to change `description` to `summary` rather than duplicating it.

**Review by pvdbosch — CHANGES_REQUESTED** [review]

File: integrationtest/src/it/externalReferencesSucceeds/openapi.yaml (hunk 31 above)
Comment (MDoevenspeck) [reply integrationtest/src/it/externalReferencesSucceeds/openapi.yaml:333]: changed and updated in rest-guide demo and rest style guide
